app/particle/particle_simulation.py

```diff
@@ -63,87 +63,112 @@
         help = "force constant")
 args = parser.parse_args()
 
-params.num_particles = args.numparticles if args.numparticles else 100
 params.simulation_height = args.height if args.height else 1000
 params.simulation_width = args.width if args.width else 1000
 params.simulation_depth = args.depth if args.depth else 1000
-params.dt = args.dt if args.dt else 0.0005
-#params.force = args.force if args.force else 100
-params.force = args.force if args.force else 100000
+
+params.num_particles = args.numparticles if args.numparticles else 100
+params.new_num_particles = params.num_particles
+params.max_radius = min(params.simulation_width, params.simulation_height, params.simulation_depth)//32
+
 params.num_active_workers = 0
 params.new_num_active_workers = 0
 params.partitions = {}
-params.max_radius = min(params.simulation_width, params.simulation_height, params.simulation_depth)//32
+
+params.dt = args.dt if args.dt else 0.0005
 params.timesteps_per_second = 0
+params.force = args.force if args.force else 100000
+
 params.init_total_energy = 0.0
 params.curr_total_energy = 0.0
 
+def add_one_particle(particle_id):
+    radius = 30
+    position = [random.randint(radius, params.simulation_width - 1),
+                random.randint(radius, params.simulation_height - 1),
+                random.randint(radius, params.simulation_depth - 1)]
+    velocity = [400*random.randint(0,radius//4),
+                400*random.randint(0,radius//4),
+                400*random.randint(0,radius//4)]
+    mass = 3
+    params.init_total_energy += 0.5 * mass * (velocity[0]**2 + velocity[1]**2 + velocity[2]**2)
+    thread_num = util.determine_particle_thread_num(position[0])
+    new_particle = Particle(particle_id, thread_num, position, velocity, mass, radius)
+    params.partitions[thread_num].add_particle(new_particle)
+
+def subtract_one_particle(particle_id):
+    if params.rank is 0:
+        for i in range(1, params.num_active_workers+1):
+            new_particles = params.comm.recv(source = mpi.ANY_SOURCE, status = params.mpi_status, tag = 0)
+            params.partitions[params.mpi_status.Get_source()].particles = new_particles
+        params.init_total_energy -= params.comm.recv(source = mpi.ANY_SOURCE, status = params.mpi_status, tag = 88)
+    elif params.rank <= params.num_active_workers:
+        partition = params.partitions[params.rank]
+        particles = partition.particles
+        for particle in particles:
+            if particle.particle_id == particle_id:
+                params.comm.send(particle.particle_initial_energy, tag = 88)
+                particles.remove(particle)
+                break
```

Comment:
I doubt there's time for you to change this now, but looping over every particle is much slower than looking up the correct particle in a list (preferable) or dict (a bit slower).  Keep in mind that the efficiency of code matters in general, but especially in parallel computing:  every node is going to be blocking on this loop.